# Project Title
### Data Engineering Capstone Project

#### Project Summary
--describe your project at a high level--

The project follows the follow steps:
* Step 1: Scope the Project and Gather Data
* Step 2: Explore and Assess the Data
* Step 3: Define the Data Model
* Step 4: Run ETL to Model the Data
* Step 5: Complete Project Write Up

In [1]:
# Do all imports and installs here
import pandas as pd

### Step 1: Scope the Project and Gather Data

#### Scope 
Explain what you plan to do in the project in more detail. What data do you use? What is your end solution look like? What tools did you use? etc>

#### Describe and Gather Data 
Describe the data sets you're using. Where did it come from? What type of information is included? 

In [2]:
# Read in the data here

In [3]:
from pyspark.sql import SparkSession

# config("spark.jars.packages", "saurfang:spark-sas7bdat:2.0.0-s_2.11").\

spark = SparkSession.builder.\
config("spark.jars.repositories", "https://repos.spark-packages.org/").\
config("spark.jars.packages", "saurfang:spark-sas7bdat:3.0.0-s_2.12").\
enableHiveSupport().getOrCreate()

# df_spark = spark.read.format('com.github.saurfang.sas.spark').load('data/18-83510-I94-Data-2016/i94_apr16_sub.sas7bdat')


21/10/23 21:35:02 WARN Utils: Your hostname, lwo1-lhp-a05809 resolves to a loopback address: 127.0.1.1; using 192.168.0.201 instead (on interface wlp0s20f3)
21/10/23 21:35:02 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
https://repos.spark-packages.org/ added as a remote repository with the name: repo-1


:: loading settings :: url = jar:file:/home/roman/Documents/spark-3.1.2-bin-hadoop3.2/jars/ivy-2.4.0.jar!/org/apache/ivy/core/settings/ivysettings.xml


Ivy Default Cache set to: /home/roman/.ivy2/cache
The jars for the packages stored in: /home/roman/.ivy2/jars
saurfang#spark-sas7bdat added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-f08d87e8-80c7-44b4-8b58-6fdbb7cca6d2;1.0
	confs: [default]
	found saurfang#spark-sas7bdat;3.0.0-s_2.12 in spark-packages
	found com.epam#parso;2.0.11 in central
	found org.slf4j#slf4j-api;1.7.5 in local-m2-cache
	found org.apache.logging.log4j#log4j-api-scala_2.12;12.0 in central
	found org.scala-lang#scala-reflect;2.12.10 in user-list
	found org.apache.logging.log4j#log4j-api;2.13.2 in central
:: resolution report :: resolve 2557ms :: artifacts dl 14ms
	:: modules in use:
	com.epam#parso;2.0.11 from central in [default]
	org.apache.logging.log4j#log4j-api;2.13.2 from central in [default]
	org.apache.logging.log4j#log4j-api-scala_2.12;12.0 from central in [default]
	org.scala-lang#scala-reflect;2.12.10 from user-list in [default]
	org.slf4j#slf4j-api;1.7.5 from local-

In [4]:
#write to parquet
# df_spark.write.mode("overwrite").parquet("sas_data")
df_spark=spark.read.parquet("sas_data")

### Step 2: Explore and Assess the Data
#### Explore the Data 
Identify data quality issues, like missing values, duplicate data, etc.

#### Cleaning Steps
Document steps necessary to clean the data

In [5]:
df=df_spark.toPandas()
spark.stop()

21/10/23 21:35:11 WARN package: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.


In [6]:
df_spark.count()

3096313

In [6]:
df.head()

cicid   i94yr  i94mon  i94cit  i94res i94port  arrdate  i94mode  \
0  5749531.0  2016.0     4.0   251.0   251.0     NYC  20574.0      1.0   
1  5749532.0  2016.0     4.0   251.0   251.0     NYC  20574.0      1.0   
2  5749533.0  2016.0     4.0   251.0   251.0     NYC  20574.0      1.0   
3  5749538.0  2016.0     4.0   251.0   251.0     NYC  20574.0      1.0   
4  5749540.0  2016.0     4.0   251.0   251.0     NYC  20574.0      1.0   

  i94addr  depdate  ...  entdepu  matflag  biryear   dtaddto gender insnum  \
0      NV  20579.0  ...     None        M   1974.0  10292016      F   None   
1      NV  20584.0  ...     None        M   1976.0  10292016      F   None   
2      NV  20584.0  ...     None        M   1977.0  10292016      M   None   
3      NY      NaN  ...     None     None   1993.0  10292016      M   None   
4      NY  20575.0  ...     None        M   1977.0  10292016      M   None   

  airline        admnum  fltno visatype  
0      DL  9.493189e+10  00469       B1  
1      DL  9.493207e+10  00469       B2  
2      DL  9.493198e+10  00469       B2  
3      EK  9.495135e+10  00203       B2  
4      PR  9.502526e+10  00126       B2  

[5 rows x 28 columns]

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3096313 entries, 0 to 3096312
Data columns (total 28 columns):
 #   Column    Dtype  
---  ------    -----  
 0   cicid     float64
 1   i94yr     float64
 2   i94mon    float64
 3   i94cit    float64
 4   i94res    float64
 5   i94port   object 
 6   arrdate   float64
 7   i94mode   float64
 8   i94addr   object 
 9   depdate   float64
 10  i94bir    float64
 11  i94visa   float64
 12  count     float64
 13  dtadfile  object 
 14  visapost  object 
 15  occup     object 
 16  entdepa   object 
 17  entdepd   object 
 18  entdepu   object 
 19  matflag   object 
 20  biryear   float64
 21  dtaddto   object 
 22  gender    object 
 23  insnum    object 
 24  airline   object 
 25  admnum    float64
 26  fltno     object 
 27  visatype  object 
dtypes: float64(13), object(15)
memory usage: 661.4+ MB


In [10]:
print(df.shape)

(3096313, 28)


In [11]:
df.describe(include='all')

cicid      i94yr     i94mon        i94cit        i94res  \
count   3.096313e+06  3096313.0  3096313.0  3.096313e+06  3.096313e+06   
unique           NaN        NaN        NaN           NaN           NaN   
top              NaN        NaN        NaN           NaN           NaN   
freq             NaN        NaN        NaN           NaN           NaN   
mean    3.078652e+06     2016.0        4.0  3.049069e+02  3.032838e+02   
std     1.763278e+06        0.0        0.0  2.100269e+02  2.085832e+02   
min     6.000000e+00     2016.0        4.0  1.010000e+02  1.010000e+02   
25%     1.577790e+06     2016.0        4.0  1.350000e+02  1.310000e+02   
50%     3.103507e+06     2016.0        4.0  2.130000e+02  2.130000e+02   
75%     4.654341e+06     2016.0        4.0  5.120000e+02  5.040000e+02   
max     6.102785e+06     2016.0        4.0  9.990000e+02  7.600000e+02   

        i94port       arrdate       i94mode  i94addr       depdate  ...  \
count   3096313  3.096313e+06  3.096074e+06  2943721  2.953856e+06  ...   
unique      299           NaN           NaN      457           NaN  ...   
top         NYC           NaN           NaN       FL           NaN  ...   
freq     485916           NaN           NaN   621701           NaN  ...   
mean        NaN  2.055985e+04  1.073690e+00      NaN  2.057395e+04  ...   
std         NaN  8.777339e+00  5.158963e-01      NaN  2.935697e+01  ...   
min         NaN  2.054500e+04  1.000000e+00      NaN  1.517600e+04  ...   
25%         NaN  2.055200e+04  1.000000e+00      NaN  2.056100e+04  ...   
50%         NaN  2.056000e+04  1.000000e+00      NaN  2.057000e+04  ...   
75%         NaN  2.056700e+04  1.000000e+00      NaN  2.057900e+04  ...   
max         NaN  2.057400e+04  9.000000e+00      NaN  4.542700e+04  ...   

        entdepu  matflag       biryear   dtaddto   gender  insnum  airline  \
count       392  2957884  3.095511e+06   3095836  2682044  113708  3012686   
unique        2        1           NaN       777        4    1913      534   
top           U        M           NaN  07282016        M    3692       AA   
freq        391  2957884           NaN     67889  1377224    2155   310091   
mean        NaN      NaN  1.974232e+03       NaN      NaN     NaN      NaN   
std         NaN      NaN  1.742026e+01       NaN      NaN     NaN      NaN   
min         NaN      NaN  1.902000e+03       NaN      NaN     NaN      NaN   
25%         NaN      NaN  1.962000e+03       NaN      NaN     NaN      NaN   
50%         NaN      NaN  1.975000e+03       NaN      NaN     NaN      NaN   
75%         NaN      NaN  1.986000e+03       NaN      NaN     NaN      NaN   
max         NaN      NaN  2.019000e+03       NaN      NaN     NaN      NaN   

              admnum    fltno visatype  
count   3.096313e+06  3076764  3096313  
unique           NaN     7152       17  
top              NaN     LAND       WT  
freq             NaN    44297  1309059  
mean    7.082885e+10      NaN      NaN  
std     2.215442e+10      NaN      NaN  
min     0.000000e+00      NaN      NaN  
25%     5.603523e+10      NaN      NaN  
50%     5.936094e+10      NaN      NaN  
75%     9.350987e+10      NaN      NaN  
max     9.991557e+10      NaN      NaN  

[11 rows x 28 columns]

In [12]:
def data_inv(df):
    print('number of records: ',df.shape[0])
    print('columns count: ',df.shape[1])
    print('-'*10)
    print('dateset columns: \n')
    print(df.columns)
    print('-'*10)
    print('data-type of each column: \n')
    print(df.dtypes)
    print('-'*10)
    print('missing rows in each column: \n')
    c=df.isnull().sum()
    print(c[c>0])
data_inv(df)

netflix movies and shows:  3096313
dataset variables:  28
----------
dateset columns: 

Index(['cicid', 'i94yr', 'i94mon', 'i94cit', 'i94res', 'i94port', 'arrdate',
       'i94mode', 'i94addr', 'depdate', 'i94bir', 'i94visa', 'count',
       'dtadfile', 'visapost', 'occup', 'entdepa', 'entdepd', 'entdepu',
       'matflag', 'biryear', 'dtaddto', 'gender', 'insnum', 'airline',
       'admnum', 'fltno', 'visatype'],
      dtype='object')
----------
data-type of each column: 

cicid       float64
i94yr       float64
i94mon      float64
i94cit      float64
i94res      float64
i94port      object
arrdate     float64
i94mode     float64
i94addr      object
depdate     float64
i94bir      float64
i94visa     float64
count       float64
dtadfile     object
visapost     object
occup        object
entdepa      object
entdepd      object
entdepu      object
matflag      object
biryear     float64
dtaddto      object
gender       object
insnum       object
airline      object
admnum      float64
f

## Data dictionary
Column|Description| Key 
------|----|-----
I94YR| 4 digit year 
I94MON| Numeric month 
I94CIT & I94RES| This format shows all the valid and invalid codes for processing | 582 =  'MEXICO Air Sea, and Not Reported (I-94, no land arrivals)'<br>   236 =  'AFGHANISTAN'<br>   101 =  'ALBANIA'<br>   316 =  'ALGERIA'<br>   102 =  'ANDORRA'<br>   324 =  'ANGOLA'<br>   529 =  'ANGUILLA'<br>   518 =  'ANTIGUA-BARBUDA'<br>   687 =  'ARGENTINA '<br>   151 =  'ARMENIA'<br>   532 =  'ARUBA'<br>   438 =  'AUSTRALIA'<br>   103 =  'AUSTRIA'<br>   152 =  'AZERBAIJAN'<br>   512 =  'BAHAMAS'<br>   298 =  'BAHRAIN'<br>   274 =  'BANGLADESH'<br>   513 =  'BARBADOS'<br>   104 =  'BELGIUM'<br>   581 =  'BELIZE'<br>   386 =  'BENIN'<br>   509 =  'BERMUDA'<br>   153 =  'BELARUS'<br>   242 =  'BHUTAN'<br>   688 =  'BOLIVIA'<br>   717 =  'BONAIRE, ST EUSTATIUS, SABA' <br>   164 =  'BOSNIA-HERZEGOVINA'<br>   336 =  'BOTSWANA'<br>   689 =  'BRAZIL'<br>   525 =  'BRITISH VIRGIN ISLANDS'<br>   217 =  'BRUNEI'<br>   105 =  'BULGARIA'<br>   393 =  'BURKINA FASO'<br>   243 =  'BURMA'<br>   375 =  'BURUNDI'<br>   310 =  'CAMEROON'<br>   326 =  'CAPE VERDE'<br>   526 =  'CAYMAN ISLANDS'<br>   383 =  'CENTRAL AFRICAN REPUBLIC'<br>   384 =  'CHAD'<br>   690 =  'CHILE'<br>   245 =  'CHINA, PRC'<br>   721 =  'CURACAO' <br>   270 =  'CHRISTMAS ISLAND'<br>   271 =  'COCOS ISLANDS'<br>   691 =  'COLOMBIA'<br>   317 =  'COMOROS'<br>   385 =  'CONGO'<br>   467 =  'COOK ISLANDS'<br>   575 =  'COSTA RICA'<br>   165 =  'CROATIA'<br>   584 =  'CUBA'<br>   218 =  'CYPRUS'<br>   140 =  'CZECH REPUBLIC'<br>   723 =  'FAROE ISLANDS (PART OF DENMARK)'  <br>   108 =  'DENMARK'<br>   322 =  'DJIBOUTI'<br>   519 =  'DOMINICA'<br>   585 =  'DOMINICAN REPUBLIC'<br>   240 =  'EAST TIMOR'<br>   692 =  'ECUADOR'<br>   368 =  'EGYPT'<br>   576 =  'EL SALVADOR'<br>   399 =  'EQUATORIAL GUINEA'<br>   372 =  'ERITREA'<br>   109 =  'ESTONIA'<br>   369 =  'ETHIOPIA'<br>   604 =  'FALKLAND ISLANDS'<br>   413 =  'FIJI'<br>   110 =  'FINLAND'<br>   111 =  'FRANCE'<br>   601 =  'FRENCH GUIANA'<br>   411 =  'FRENCH POLYNESIA'<br>   387 =  'GABON'<br>   338 =  'GAMBIA'<br>   758 =  'GAZA STRIP' <br>   154 =  'GEORGIA'<br>   112 =  'GERMANY'<br>   339 =  'GHANA'<br>   143 =  'GIBRALTAR'<br>   113 =  'GREECE'<br>   520 =  'GRENADA'<br>   507 =  'GUADELOUPE'<br>   577 =  'GUATEMALA'<br>   382 =  'GUINEA'<br>   327 =  'GUINEA-BISSAU'<br>   603 =  'GUYANA'<br>   586 =  'HAITI'<br>   726 =  'HEARD AND MCDONALD IS.'<br>   149 =  'HOLY SEE/VATICAN'<br>   528 =  'HONDURAS'<br>   206 =  'HONG KONG'<br>   114 =  'HUNGARY'<br>   115 =  'ICELAND'<br>   213 =  'INDIA'<br>   759 =  'INDIAN OCEAN AREAS (FRENCH)' <br>   729 =  'INDIAN OCEAN TERRITORY' <br>   204 =  'INDONESIA'<br>   249 =  'IRAN'<br>   250 =  'IRAQ'<br>   116 =  'IRELAND'<br>   251 =  'ISRAEL'<br>   117 =  'ITALY'<br>   388 =  'IVORY COAST'<br>   514 =  'JAMAICA'<br>   209 =  'JAPAN'<br>   253 =  'JORDAN'<br>   201 =  'KAMPUCHEA'<br>   155 =  'KAZAKHSTAN'<br>   340 =  'KENYA'<br>   414 =  'KIRIBATI'<br>   732 =  'KOSOVO' <br>   272 =  'KUWAIT'<br>   156 =  'KYRGYZSTAN'<br>   203 =  'LAOS'<br>   118 =  'LATVIA'<br>   255 =  'LEBANON'<br>   335 =  'LESOTHO'<br>   370 =  'LIBERIA'<br>   381 =  'LIBYA'<br>   119 =  'LIECHTENSTEIN'<br>   120 =  'LITHUANIA'<br>   121 =  'LUXEMBOURG'<br>   214 =  'MACAU'<br>   167 =  'MACEDONIA'<br>   320 =  'MADAGASCAR'<br>   345 =  'MALAWI'<br>   273 =  'MALAYSIA'<br>   220 =  'MALDIVES'<br>   392 =  'MALI'<br>   145 =  'MALTA'<br>   472 =  'MARSHALL ISLANDS'<br>   511 =  'MARTINIQUE'<br>   389 =  'MAURITANIA'<br>   342 =  'MAURITIUS'<br>   760 =  'MAYOTTE (AFRICA - FRENCH)' <br>   473 =  'MICRONESIA, FED. STATES OF'<br>   157 =  'MOLDOVA'<br>   122 =  'MONACO'<br>   299 =  'MONGOLIA'<br>   735 =  'MONTENEGRO' <br>   521 =  'MONTSERRAT'<br>   332 =  'MOROCCO'<br>   329 =  'MOZAMBIQUE'<br>   371 =  'NAMIBIA'<br>   440 =  'NAURU'<br>   257 =  'NEPAL'<br>   123 =  'NETHERLANDS'<br>   508 =  'NETHERLANDS ANTILLES'<br>   409 =  'NEW CALEDONIA'<br>   464 =  'NEW ZEALAND'<br>   579 =  'NICARAGUA'<br>   390 =  'NIGER'<br>   343 =  'NIGERIA'<br>   470 =  'NIUE'<br>   275 =  'NORTH KOREA'<br>   124 =  'NORWAY'<br>   256 =  'OMAN'<br>   258 =  'PAKISTAN'<br>   474 =  'PALAU'<br>   743 =  'PALESTINE' <br>   504 =  'PANAMA'<br>   441 =  'PAPUA NEW GUINEA'<br>   693 =  'PARAGUAY'<br>   694 =  'PERU'<br>   260 =  'PHILIPPINES'<br>   416 =  'PITCAIRN ISLANDS'<br>   107 =  'POLAND'<br>   126 =  'PORTUGAL'<br>   297 =  'QATAR'<br>   748 =  'REPUBLIC OF SOUTH SUDAN'<br>   321 =  'REUNION'<br>   127 =  'ROMANIA'<br>   158 =  'RUSSIA'<br>   376 =  'RWANDA'<br>   128 =  'SAN MARINO'<br>   330 =  'SAO TOME AND PRINCIPE'<br>   261 =  'SAUDI ARABIA'<br>   391 =  'SENEGAL'<br>   142 =  'SERBIA AND MONTENEGRO'<br>   745 =  'SERBIA' <br>   347 =  'SEYCHELLES'<br>   348 =  'SIERRA LEONE'<br>   207 =  'SINGAPORE'<br>   141 =  'SLOVAKIA'<br>   166 =  'SLOVENIA'<br>   412 =  'SOLOMON ISLANDS'<br>   397 =  'SOMALIA'<br>   373 =  'SOUTH AFRICA'<br>   276 =  'SOUTH KOREA'<br>   129 =  'SPAIN'<br>   244 =  'SRI LANKA'<br>   346 =  'ST. HELENA'<br>   522 =  'ST. KITTS-NEVIS'<br>   523 =  'ST. LUCIA'<br>   502 =  'ST. PIERRE AND MIQUELON'<br>   524 =  'ST. VINCENT-GRENADINES'<br>   716 =  'SAINT BARTHELEMY' <br>   736 =  'SAINT MARTIN' <br>   749 =  'SAINT MAARTEN' <br>   350 =  'SUDAN'<br>   602 =  'SURINAME'<br>   351 =  'SWAZILAND'<br>   130 =  'SWEDEN'<br>   131 =  'SWITZERLAND'<br>   262 =  'SYRIA'<br>   268 =  'TAIWAN'<br>   159 =  'TAJIKISTAN'<br>   353 =  'TANZANIA'<br>   263 =  'THAILAND'<br>   304 =  'TOGO'<br>   417 =  'TONGA'<br>   516 =  'TRINIDAD AND TOBAGO'<br>   323 =  'TUNISIA'<br>   264 =  'TURKEY'<br>   161 =  'TURKMENISTAN'<br>   527 =  'TURKS AND CAICOS ISLANDS'<br>   420 =  'TUVALU'<br>   352 =  'UGANDA'<br>   162 =  'UKRAINE'<br>   296 =  'UNITED ARAB EMIRATES'<br>   135 =  'UNITED KINGDOM'<br>   695 =  'URUGUAY'<br>   163 =  'UZBEKISTAN'<br>   410 =  'VANUATU'<br>   696 =  'VENEZUELA'<br>   266 =  'VIETNAM'<br>   469 =  'WALLIS AND FUTUNA ISLANDS'<br>   757 =  'WEST INDIES (FRENCH)' <br>   333 =  'WESTERN SAHARA'<br>   465 =  'WESTERN SAMOA'<br>   216 =  'YEMEN'<br>   139 =  'YUGOSLAVIA'<br>   301 =  'ZAIRE'<br>   344 =  'ZAMBIA'<br>   315 =  'ZIMBABWE'<br>   403 =  'INVALID: AMERICAN SAMOA'<br>   712 =  'INVALID: ANTARCTICA' <br>   700 =  'INVALID: BORN ON BOARD SHIP'<br>   719 =  'INVALID: BOUVET ISLAND (ANTARCTICA/NORWAY TERR.)'<br>   574 =  'INVALID: CANADA'<br>   720 =  'INVALID: CANTON AND ENDERBURY ISLS' <br>   106 =  'INVALID: CZECHOSLOVAKIA'<br>   739 =  'INVALID: DRONNING MAUD LAND (ANTARCTICA-NORWAY)' <br>   394 =  'INVALID: FRENCH SOUTHERN AND ANTARCTIC'<br>   501 =  'INVALID: GREENLAND'<br>   404 =  'INVALID: GUAM'<br>   730 =  'INVALID: INTERNATIONAL WATERS' <br>   731 =  'INVALID: JOHNSON ISLAND' <br>   471 =  'INVALID: MARIANA ISLANDS, NORTHERN'<br>   737 =  'INVALID: MIDWAY ISLANDS' <br>   753 =  'INVALID: MINOR OUTLYING ISLANDS - USA'<br>   740 =  'INVALID: NEUTRAL ZONE (S. ARABIA/IRAQ)' <br>   710 =  'INVALID: NON-QUOTA IMMIGRANT'<br>   505 =  'INVALID: PUERTO RICO'<br>    0  =  'INVALID: STATELESS'<br>   705 =  'INVALID: STATELESS'<br>   583 =  'INVALID: UNITED STATES'<br>   407 =  'INVALID: UNITED STATES'<br>   999 =  'INVALID: UNKNOWN'<br>   239 =  'INVALID: UNKNOWN COUNTRY'<br>   134 =  'INVALID: USSR'<br>   506 =  'INVALID: U.S. VIRGIN ISLANDS'<br>   755 =  'INVALID: WAKE ISLAND'  <br>   311 =  'Collapsed Tanzania (should not show)'<br>   741 =  'Collapsed Curacao (should not show)'<br>    54 =  'No Country Code (54)'<br>   100 =  'No Country Code (100)'<br>   187 =  'No Country Code (187)'<br>   190 =  'No Country Code (190)'<br>   200 =  'No Country Code (200)'<br>   219 =  'No Country Code (219)'<br>   238 =  'No Country Code (238)'<br>   277 =  'No Country Code (277)'<br>   293 =  'No Country Code (293)'<br>   300 =  'No Country Code (300)'<br>   319 =  'No Country Code (319)'<br>   365 =  'No Country Code (365)'<br>   395 =  'No Country Code (395)'<br>   400 =  'No Country Code (400)'<br>   485 =  'No Country Code (485)'<br>   503 =  'No Country Code (503)'<br>   589 =  'No Country Code (589)'<br>   592 =  'No Country Code (592)'<br>   791 =  'No Country Code (791)'<br>   849 =  'No Country Code (849)'<br>   914 =  'No Country Code (914)'<br>   944 =  'No Country Code (944)'<br>   996 =  'No Country Code (996)' ;<br><br> 
I94PORT| This format shows all the valid and invalid codes for processing works for you 
ARRDATE| Arrival Date in the USA. It is a SAS date numeric field that a permament format has not been applied.  Please apply whichever date format works for you
I94MODE| There are missing values as well as not reported (9) shows what we have found to be valid, everything else goes into 'other' works for you. 
I94ADDR| There is lots of invalid codes in this variable and the list below shows what we have found to be valid, everything else goes into 'other' | 'AL'='ALABAMA'<br>   'AK'='ALASKA'<br> 'AZ'='ARIZONA'<br>   'AR'='ARKANSAS'<br>  'CA'='CALIFORNIA'<br>   'CO'='COLORADO'<br>  'CT'='CONNECTICUT'<br>  'DE'='DELAWARE'<br>  'DC'='DIST. OF COLUMBIA'<br>  'FL'='FLORIDA'<br>   'GA'='GEORGIA'<br>   'GU'='GUAM'<br>   'HI'='HAWAII'<br> 'ID'='IDAHO'<br>  'IL'='ILLINOIS'<br>  'IN'='INDIANA'<br>   'IA'='IOWA'<br>   'KS'='KANSAS'<br> 'KY'='KENTUCKY'<br>  'LA'='LOUISIANA'<br> 'ME'='MAINE'<br>  'MD'='MARYLAND'<br>  'MA'='MASSACHUSETTS'<br>   'MI'='MICHIGAN'<br>  'MN'='MINNESOTA'<br> 'MS'='MISSISSIPPI'<br>  'MO'='MISSOURI'<br>  'MT'='MONTANA'<br>   'NC'='N. CAROLINA'<br>  'ND'='N. DAKOTA'<br> 'NE'='NEBRASKA'<br>  'NV'='NEVADA'<br> 'NH'='NEW HAMPSHIRE'<br>   'NJ'='NEW JERSEY'<br>   'NM'='NEW MEXICO'<br>   'NY'='NEW YORK'<br>  'OH'='OHIO'<br>   'OK'='OKLAHOMA'<br>  'OR'='OREGON'<br> 'PA'='PENNSYLVANIA'<br> 'PR'='PUERTO RICO'<br>  'RI'='RHODE ISLAND'<br> 'SC'='S. CAROLINA'<br>  'SD'='S. DAKOTA'<br> 'TN'='TENNESSEE'<br> 'TX'='TEXAS'<br>  'UT'='UTAH'<br>   'VT'='VERMONT'<br>   'VI'='VIRGIN ISLANDS'<br>  'VA'='VIRGINIA'<br>  'WV'='W. VIRGINIA'<br>  'WA'='WASHINGTON'<br>   'WI'='WISCONSON'<br> 'WY'='WYOMING' <br>  '99'='All Other Codes'
DEPDATE| Departure Date from the USA. It is a SAS date numeric field that a permament format has not been applied.  Please apply whichever date format works for you
I94BIR| Age of Respondent in Years 
I94VISA - Visa codes collapsed into three categories| 1 = Business, 2 = Pleasure, 3 = Student
COUNT| Used for summary statistics 
DTADFILE| Character Date Field - Date added to I-94 Files - CIC does not use 
VISAPOST| Department of State where where Visa was issued - CIC does not use 
OCCUP | Occupation that will be performed in U.S. - CIC does not use 
ENTDEPA | Arrival Flag - admitted or paroled into the U.S. - CIC does not use 
ENTDEPD | Departure Flag - Departed, lost I-94 or is deceased - CIC does not use 
ENTDEPU | Update Flag - Either apprehended, overstayed, adjusted to perm residence - CIC does not use 
MATFLAG | Match flag - Match of arrival and departure records 
BIRYEAR | 4 digit year of birth 
DTADDTO | Character Date Field - Date to which admitted to U.S. (allowed to stay until) - CIC does not use 
GENDER | Non-immigrant sex 
INSNUM | INS number 
AIRLINE | Airline used to arrive in U.S. 
ADMNUM | Admission Number 
FLTNO | Flight number of Airline used to arrive in U.S. 
VISATYPE | Class of admission legally admitting the non-immigrant to temporarily stay in U.S. 


### Step 3: Define the Data Model
#### 3.1 Conceptual Data Model
Map out the conceptual data model and explain why you chose that model

#### 3.2 Mapping Out Data Pipelines
List the steps necessary to pipeline the data into the chosen data model

### Step 4: Run Pipelines to Model the Data 
#### 4.1 Create the data model
Build the data pipelines to create the data model.

In [ ]:
# Write code here

#### 4.2 Data Quality Checks
Explain the data quality checks you'll perform to ensure the pipeline ran as expected. These could include:
 * Integrity constraints on the relational database (e.g., unique key, data type, etc.)
 * Unit tests for the scripts to ensure they are doing the right thing
 * Source/Count checks to ensure completeness
 
Run Quality Checks

In [ ]:
# Perform quality checks here

#### 4.3 Data dictionary 
Create a data dictionary for your data model. For each field, provide a brief description of what the data is and where it came from. You can include the data dictionary in the notebook or in a separate file.

#### Step 5: Complete Project Write Up
* Clearly state the rationale for the choice of tools and technologies for the project.
* Propose how often the data should be updated and why.
* Write a description of how you would approach the problem differently under the following scenarios:
 * The data was increased by 100x.
 * The data populates a dashboard that must be updated on a daily basis by 7am every day.
 * The database needed to be accessed by 100+ people.